src/storage/src/hummock/upload_handle_manager.rs

```diff
@@ -0,0 +1,354 @@
+        loop {
+            let mut select_all_fut = if let Some(select_all_fut) = self.select_all.take() {
+                select_all_fut
+            } else {
+                assert!(self.manager.remaining_handle_count.is_empty());
+                // If `self.select_all` is `None`, there is no pending epoch. Return `Pending`.
+                return Poll::Pending;
+            };
+            match select_all_fut.poll_unpin(cx) {
+                Poll::Ready(((result, epoch), _, futures)) => {
+                    // Reset `self.select_all`
+                    if !futures.is_empty() {
+                        assert!(self.select_all.replace(select_all(futures)).is_none());
+                    }
+
+                    // Decrease the remaining count of the epoch by 1 and remove it when it reaches
+                    // 0.
+                    let epoch_remaining_count_mut_ref =
+                        self.manager.remaining_handle_count.get_mut(&epoch).expect(
+                            "a join handle just finish. prev count must not zero. should exist",
+                        );
+                    assert!(*epoch_remaining_count_mut_ref > 0);
+                    *epoch_remaining_count_mut_ref -= 1;
+                    let epoch_remaining_count = *epoch_remaining_count_mut_ref;
+                    if epoch_remaining_count == 0 {
+                        let _ = self.manager.remaining_handle_count.remove(&epoch);
+                    } else {
+                        assert!(
+                            self.select_all.is_some(),
+                            "an epoch has some remaining join handle, the select_all must not be empty"
+                        );
+                    }
+
+                    match result {
+                        Ok(_) => {
+                            // If the there is no remaining join handle in this epoch, return the
+                            // epoch. Otherwise, keep polling other join handle
+                            if epoch_remaining_count == 0 {
+                                return Poll::Ready(Ok(epoch));
+                            }
+                        }
+                        Err(e) => {
+                            // Return when we meet an error
+                            return Poll::Ready(Err(e));
+                        }
+                    };
+                }
+                Poll::Pending => {
+                    assert!(self.select_all.replace(select_all_fut).is_none());
+                    return Poll::Pending;
+                }
+            };
+        }
+    }
+}
+
+impl<'a> Drop for UploadHandleManagerNextFinishedEpoch<'a> {
+    fn drop(&mut self) {
+        // In case of being dropped, restore the join handle back to `self.manager`
+        if let Some(select_all) = self.select_all.take() {
+            self.manager
+                .epoch_upload_handle
+                .extend(select_all.into_inner());
```

Comment:
This will cause the order of epoch upload handle to be different from the flush order. Will this be an issue?

Personally I think it is fine because we anyway select all on next_finished_epoch but still want to double confirm.

Reply:
Yes, this will not be an issue since each upload task is independent to each other. 

Moreover, what stores in `select_all` is `JoinHandle` of concurrent upload tasks. Here we just reorder joining the `JoinHandle`. The task execution order is the same as the main branch, which is handled by tokio runtime.